pkg/controller/apiextensions/composite/api.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package composite
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+)
+
+// Error strings.
+const (
+	errApplySecret = "cannot apply connection secret"
+)
+
+// APIFilteredSecretPublisher publishes ConnectionDetails content after filtering
+// it through a set of permitted keys.
+type APIFilteredSecretPublisher struct {
+	client resource.Applicator
+	filter []string
+}
+
+// NewAPIFilteredSecretPublisher returns a new APIFilteredSecretPublisher.
+func NewAPIFilteredSecretPublisher(c client.Client, filter []string) *APIFilteredSecretPublisher {
+	// NOTE(negz): We transparently inject an APIPatchingApplicator in order to maintain
+	// backward compatibility with the original API of this function.
+	return &APIFilteredSecretPublisher{client: resource.NewAPIPatchingApplicator(c), filter: filter}
```

Comment:
Actually, what we need is precisely `APIPatchingApplicatior` rather than any `resource.Applicatior` because the logic in the function relies on patching for partial key submission, i.e. `UpdateApplicator` would break its logic.

Reply:
I'm sure that's true, but I still think it's better to accept an interface we can stub out for testing, rather than testing through to the implementation details of the patching applicator.